### PR TITLE
Tile depth

### PR DESCRIPTION
**Pull request by michaelJwilson**
I think everything is in order so here's a PR, in case you want it for next week.  Happy to hear further suggestions.

### CODE REVIEW

File: py/desispec/scripts/master_nea.py
Comment (sbailey) [inline py/desispec/scripts/master_nea.py:20]: Are the NEA files sensitive to the exact details of the wavelength grid?  This is likely to change again in the future, e.g. to extend the wavelength range further blue.  If it isn't sensitive to the sampling but perhaps is sensitive to the edges (e.g. due to being able to interpolate but not extrapolate) we should make sure this is a superset of what we might need in the future, e.g. 3520 - 9950.

File: py/desispec/scripts/master_nea.py
Comment (sbailey) [inline py/desispec/scripts/master_nea.py:35]: please add docstring

File: py/desispec/scripts/master_nea.py
Comment (sbailey) [inline py/desispec/scripts/master_nea.py:56]: Is this try/except needed, or leftover from testing.  If still needed, let's use the logger instead of a bare print.

File: py/desispec/scripts/procexp.py
Comment (sbailey) [inline py/desispec/scripts/procexp.py:2]: Something has gone wrong here.  This script originally was about processing the exposure to apply all of the calibration steps and write the final cframe files.  The comment here implies that it now only computes the template SNR, and many of the other imports have been removed, e.g. `from desispec.fluxcalibration import apply_flux_calibration`, but downstream in this code `apply_flux_calibration` is still called.  i.e. this code will not run in the pipeline as-is.

File: py/desispec/scripts/procexp.py
Comment (sbailey) [inline py/desispec/scripts/procexp.py:27]: typo that prevents this script from even being imported; i.e. this branch needs more testing.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:121]: What is the case where this SPCALIB replacement is needed?  I recognize that from header value shortening, but findfile should find the full correct path to the flux calibration file without a 'SPCALIB' in it, and `calibdir = os.environ['DESI_SPECTRO_CALIB']` only has a default fluxcalib used by nightwatch, not the nightly fluxcalib files for this night, expid.  Since `calib` shouldn't have 'SPCALIB` in it, this is a no-op, but could lead to confusion about what the code is trying to do.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:126]: This case of replacement makes sense to me, since if the night didn't get good fiberflats, it falls back to default fiberflats in desi_spectro_calib.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:134]: these hardcoded paths appears to be unused leftovers.  Please remove.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:197]: This hardcoded path, however, appears to be used.  Please scrub all hardcoded paths from this code and use options like `--outdir` instead.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:39]: All of the code from here on down in this file should be wrapped in a function so that `import desispec.scripts.tsnr_afterburner` doesn't trigger it trying to run a bunch of code.  That could happen e.g. by a test runner looking for tests, or a documentation scraper looking for every function to auto-generate docs.  It's ok to have
```
if __name__ == '__main__':
    main()
```
so that this script can be called with `python tsnr_afterburner.py` without putting in the default installed $PATH.

File: py/desispec/scripts/tsnr_afterburner.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_afterburner.py:45]: I think this is used to replace `SPCALIB` in headers with a full file path, but in that case the headers themselves should also have kept the value of `$DESI_SPECTRO_CALIB` in some DEPNAMnn/DEPVERnn keyword pair.  `desiutil.depend.getdep(header, 'DESI_SPECTRO_CALIB')` parses those for you to get the DESI_SPECTRO_CALIB value.   Using that would prevent the case of having the file generated pointing to one version of `$DESI_SEPCTRO_CALIB`.

File: py/desispec/scripts/tsnr_ensemble.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_ensemble.py:54]: please add a docstring defining inputs and e.g. what options for "tracer" are allowed, and noting that the act of creating this object writes a file to disk.

File: py/desispec/scripts/tsnr_ensemble.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_ensemble.py:55]: please add docstring

File: py/desispec/scripts/tsnr_ensemble.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_ensemble.py:57]: The default `template_ensemble` constructor `tracer` is "ELG" (uppercase), while this `tracer_maker` function appears to assume all tracers are lowercase.  Let's be consistent throughout, or otherwise include code like `tracer = tracer.lower()` to make it input case agnostic.

File: py/desispec/scripts/tsnr_ensemble.py
Comment (sbailey) [inline py/desispec/scripts/tsnr_ensemble.py:147]: This is good, and is the method I was suggesting for `tsnr_afterburner.py` too.

File: py/desispec/tsnr.py
Comment (sbailey) [inline py/desispec/tsnr.py:166]: Please expand docstring to define all inputs and what is returned (or written?)

**Review by sbailey — CHANGES_REQUESTED** [review]
Major (but hopefully easy to fix): procexp.py is broken for the spectro pipeline.  See comments inline.

Plus a bunch of cleanup and maintainability requests like removing hardcoded personal paths and adding docstrings.

File: py/desispec/scripts/master_nea.py
Comment (michaelJwilson) [reply py/desispec/scripts/master_nea.py:20]: I'm requesting the wavelengths in the current spectro-perf. reductions.  I can update the nea files to the larger range, but I'm already seeing some breakdown at the current range and patching those with a median.  We argued the wavelength dependence was smooth so extrapolation may not be so bad.

File: py/desispec/scripts/master_nea.py
Comment (michaelJwilson) [reply py/desispec/scripts/master_nea.py:56]: This is catching the un-normalized psf values as erroneous and patching.  Will change to logger.

File: py/desispec/scripts/procexp.py
Comment (michaelJwilson) [reply py/desispec/scripts/procexp.py:2]: Yes, agreed.  Sorry this has crept in by accident.  It should be a feature added to procexp, and was running as such recently.  I'll fix that up.

**Review by sbailey — APPROVED** [review]
Looks good, thanks.  I have some minor cleanup changes in my copy, but from Slack it sounds like @michaelJwilson you have some other proposed algorithmic adjustments to make, so I'll hold off on pushing my changes until we sort out potential merge conflicts.

**Comment by sbailey** [comment]
I made a bunch of changes; example outputs in /global/cfs/cdirs/desi/users/sjbailey/tsnr/blanc-tsnr, from
```
cd /global/cfs/cdirs/desi/users/sjbailey/tsnr/blanc-tsnr
desi_tsnr_afterburner --prod /global/cfs/cdirs/desi/spectro/redux/blanc \
    --night 20201215 --cameras b0,r0,z0,b1,r1,z1,b2,r2,z2 \
    --details-dir . --outfile tsnr-blanc-20201215.fits
```

Punting back to Julien for whether these outputs look like the way he'd like to use them.

A few changes to note:
  * columns / HDUS now have "TSNR2" to emphasize that these are squared quantities
  * you now have to give the name of the summary file via `--outfile`, which will enable running different nights in parallel producing different summaries
  * The summary file now has one HDU for per-frame medians (TSNR2_FRAME) and another HDU for per-expid summed over cameras and averaged over petals (TSNR2_EXPID).  Also TSNR2_EXPID is derivable from TSNR2_FRAME, it is a primary end-user "what S/N have we achieved on this tile?" result and a bit tedious to calculate.
  * I renamed `--outdir` to `--details-dir` and dropped the 'tsnr/' forced prefix to give more flexibility.  I'm not sure I like the name, but I didn't want to confuse with `--outfile` which I view as the more primary output than the per-frame quantities that going forward are also going to be included in the frame SCORES table HDU anyway.

**Comment by julienguy** [comment]
Looks good. I tested this on the night 20201216.
```
desi_tsnr_afterburner --night 20201216 --outfile summary.fits --details-dir . --prod /global/cfs/cdirs/desi/spectro/redux/blanc/
```
Some plots for the record.
Histograms of TSNR ( sqrt(TSNR2) ):
tsnr-20201216

As a function of exposure number
tsnr-vs-expid-20201216

As a function of camera, per target class
tsnr-ELG-vs-camera-20201216
tsnr-LRG-vs-camera-20201216
tsnr-QSO-vs-camera-20201216


Merging now.

**Pull request merged**
